design/proposals/proxy-support/proxy-support.md

```diff
@@ -0,0 +1,157 @@
+# Maintaining Requester IP Addresses <!-- omit in toc --> 
+
+When Conjur (or DAP) are run behind a load balancer, users are unable to use CIDR restrictions on Hosts or Layers. Furthermore, Load Balancer IP address appears in Audit logs instead of the true client IP address. This is because Conjur references the last entry in the `X-Forwarded-For` header.
+
+This decision was conscious, and meant to prevent IP spoofing. We have reached a point where customers need a mechanism to add known proxies to Conjur/DAP to identify the true identity of the requesting client. 
+
+This design document is meant to provide a possible solution to the challenge of identifying the client IP while preventing IP spoofing.
+
+- [Current Request Flow](#current-request-flow)
+- [Proposed Solution (Summary)](#proposed-solution-summary)
+- [Proposed Solution (Detailed)](#proposed-solution-detailed)
+  - [Adding Proxies with Policy](#adding-proxies-with-policy)
+  - [Limitations](#limitations)
+    - [Proxies cannot be layered](#proxies-cannot-be-layered)
+- [Rejected Alternatives](#rejected-alternatives)
+  - [Configuration using Environment Variables](#configuration-using-environment-variables)
+
+## Current Request Flow
+
+The following visualization offers a typical experience as a requests moves through a Layer 7 load balancers.
+
+![](/design/diagrams/out/proxy-support-overview/proxy-support-overview.png)
+
+As we can see, each load balancer adds its IP address to the `X-Forwarded-For` header.
+
+With our current implementation:
+```ruby
+# Rack::Request
+def ip
+  if addr = @env['HTTP_X_FORWARDED_FOR']
+    addr.split(',').last.strip
+  else
+    @env['REMOTE_ADDR']
+  end
+end
+```
+the IP address of the last load balancer (`10.2.0.1`) would be used as the request IP.
+
+## Proposed Solution (Summary)
+Users can define the proxies in front of a node(s) using Conjur Policy. This enables Conjur to use the first non-proxy IP as the client IP.
+
+## Proposed Solution (Detailed)
+
+### Adding Proxies with Policy
+
+Given the following setup
+![](/design/diagrams/out/proxy-master-node/proxy-master-node.png)
+
+we can apply the following policy:
+```yaml
+- !policy
+  id: conjur/settings/proxies
+  body:
+    - !host 
+      id: dap-master.mycompany.com
+      annotations:
+        proxy/ip-addresses: ['1.2.3.4']
+```
+
+With the above proxied request, Conjur will:
+1. Check if the `X-Forwarded-For` header has any values. In this example it will have the following string: `10.0.0.1, 1.2.3.4`. 
+2. Perform a Regex match using the host ID in the `conjur/settings/proxies` namespace against the request hostname. In this example, we'd match on `dap-master.mycompany.com`.
+3. Grab the array of IP addresses stored in the `proxy/ip-addresses` annotation (`['1.2.3.4']`).
```

Comment:
WDYT to use `restricted_to` instead of annotation ?

Reply:
How about `restricted_proxies`?  `restricted_to` is host-specific and denotes a limited pool of IPs the request can originate from.  In this case, we're looking to remove trusted proxies to allow the true requesting client identity to come through.

Reply:
@jvanderhoof  WDYT on the following alternative:
Instead of a configure the permitted proxies ips for each follower name,
We will define it generally by the proxy range with a list of ips permitted in the XFF header (restricted_proxies), and it will look like:

```
{
- !policy
  id: conjur/settings/proxies
  body:
    - !proxy
      restricted_to: [155.0.6.6\32]
      restricted_proxies: [1.0.0.6\32, 2.0.0.6\32]
}
```

In the above example only 1 proxy can access directly to the follower (155.0.6.6) and we will validate the XFF base on the restricted_proxies list

I also add an diagram that explains the flow:
![image](https://user-images.githubusercontent.com/30017008/85383352-86ad5280-b548-11ea-8bac-5d8c84a4a8a6.png)